src/cgutils.cpp

```diff
@@ -3481,23 +3481,27 @@ static void emit_cpointercheck(jl_codectx_t &ctx, const jl_cgval_t &x, const Twi
 // allocation for known size object
 // returns a prjlvalue
 static Value *emit_allocobj(jl_codectx_t &ctx, size_t static_size, Value *jt,
-                             unsigned align=sizeof(void*)) // Allocations are at least pointer alingned
+                            bool fully_initialized, unsigned align)
 {
     ++EmittedAllocObjs;
     Value *current_task = get_current_task(ctx);
     Function *F = prepare_call(jl_alloc_obj_func);
     auto call = ctx.builder.CreateCall(F, {current_task, ConstantInt::get(ctx.types().T_size, static_size), maybe_decay_untracked(ctx, jt)});
     call->setAttributes(F->getAttributes());
     if (static_size > 0)
-        call->addRetAttr(Attribute::getWithDereferenceableBytes(ctx.builder.getContext(), static_size));
-    call->addRetAttr(Attribute::getWithAlignment(ctx.builder.getContext(), Align(align)));
+        call->addRetAttr(Attribute::getWithDereferenceableBytes(call->getContext(), static_size));
+    call->addRetAttr(Attribute::getWithAlignment(call->getContext(), Align(align)));
+#if JL_LLVM_VERSION >= 150000
+    if (fully_initialized)
```

Comment:
We do that most places, but not everywhere. It could be an interesting alternative to try out on a case-by-case basis.

Reply:
Is the goal basically to hide the UB from LLVM, and let it only appear at runtime if the user does observe the memory?

Reply:
No, there is no UB with this PR at runtime, as there is no way to observe the undefined memory. I don't want to materialize a store of freeze when it does not need to be present (as that may turn into a store of zeros if LLVM can't eliminate it, and I would like to avoid that). The primary definition (uninitialized but defined) was better already at modeling the intended behavior.

Reply:
I'm not sure I follow your rationale.
If I understood this. We now, if we're not sure that the memory will be written to before being read. Will emit a freeze store to it. But the allocation is still very much undef in all cases, it's just that after either the freeze or a normal store it's fine.

Reply:
The default option is that the state is unknown and unpredictable. That is very different from the state being undefined.

Reply:
I guess I might be missing something but, why only add the attribute when we know we will fully_initialize it?

Reply:
If it is fully_initialized then the content can be assumed to be undefined behavior to observe

Reply:
Right, but if we somehow observe it in the non fully_initialized it's still UB no? (We should always look at the frozen value) but that is not the behaviour of the intrinsic on it's own

Reply:
Right. That is why we require here that every non-padding byte will be written to before the object can be observed if we are willing to inform llvm that the content started uninitialized. We could achieve an equivalent effect by adding a lifetime start call afterwards for this case (it is what we do later if we change this to a stack allocation), but adding the attribute is simpler in this case